backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/util/FreinetAgenciesLoader.kt

```diff
@@ -0,0 +1,85 @@
+import io.ktor.client.request.request
+import io.ktor.client.statement.bodyAsText
+import io.ktor.http.HttpMethod
+import io.ktor.http.HttpStatusCode
+import io.ktor.http.URLProtocol
+import io.ktor.http.path
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+
+class FreinetAgenciesLoader {
+    private val logger = LoggerFactory.getLogger(FreinetAgenciesLoader::class.java)
+    private val httpClient = HttpClient {
+        install(HttpRequestRetry) {
+            retryOnServerErrors(maxRetries = 3)
+            retryOnException(maxRetries = 3, retryOnTimeout = false)
+            exponentialDelay()
+        }
+        HttpResponseValidator {
+            validateResponse { response ->
+                if (response.status != HttpStatusCode.OK) {
+                    throw Exception(response.status.toString())
+                }
+            }
+        }
+    }
+
+    fun loadAgenciesFromXml(projectConfigs: List<ProjectConfig>): List<FreinetApiAgency> {
+        val bayernConfig =
+            projectConfigs.find { it.id == EAK_BAYERN_PROJECT } ?: throw NotFoundException("Project config not found")
+        if (bayernConfig.freinet == null) {
+            logger.error("Couldn't find required freinet api parameters in backend config.")
+            return emptyList()
+        }
+        val freinetConfig = bayernConfig.freinet
+        try {
+            val response = runBlocking {
+                httpClient.request {
+                    url {
+                        protocol = URLProtocol.HTTP
+                        host = freinetConfig.host
+                        path(freinetConfig.path)
+                        parameters.append("accessKey", freinetConfig.accessToken)
+                        parameters.append("portalId", freinetConfig.portalId)
+                        parameters.append("limit", "1000")
+                    }
+                    method = HttpMethod.Get
+                }.bodyAsText()
+            }
+
+            val xmlMapper = XmlMapper()
+            xmlMapper.registerModule(KotlinModule.Builder().build())
+            xmlMapper.enable(DeserializationFeature.ACCEPT_EMPTY_STRING_AS_NULL_OBJECT)
+            xmlMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+
+            return transformAndFilterAgencyData(xmlMapper.readValue(response, XMLAgencies::class.java))
+        } catch (e: Exception) {
+            logger.error("Couldn't fetch agency information: ", e)
+            return emptyList()
+        }
+    }
+
+    private fun transformAndFilterAgencyData(agencies: XMLAgencies): List<FreinetApiAgency> {
+        return agencies.agencies.mapNotNull {
+            FreinetApiAgency(
+                agencyId = it.agencyId!!.toInt(),
+                agencyName = it.agencyName!!,
+                apiAccessKey = it.accessKey!!,
+                officialRegionalKeys = it.officialRegionalKeys?.split(",") ?: emptyList()
+            )
```

Comment:
:x: Sorry, maybe my previous comment was misunderstandable. This does not really work/will crash if a property is null since you never check whether the properties are null and just use a non-null assertion (`!!`) (even though it seems like nothing is actually null).
`mapNotNull` only filters out `null` after mapping, therefore it does act exactly like `map` in this case. You could do the following (if changing `var` to `val` as mentioned above):
```suggestion
        return agencies.agencies.mapNotNull {
            return@mapNotNull if (it.agencyId != null && it.agencyName != null && it.accessKey != null) {
                FreinetApiAgency(
                    agencyId = it.agencyId.toInt(),
                    agencyName = it.agencyName,
                    apiAccessKey = it.accessKey,
                    officialRegionalKeys = it.officialRegionalKeys?.split(",") ?: emptyList()
                )
            } else null
```

Reply:
Just takes some time to get used to kotlin magic I guess :D